### PR TITLE
Replace 'show-ref' command with 'branch' and 'tag'

### PR DESCRIPTION
**Pull request by skyline75489**
Supports #14180.

I'm no expert at git but I found that `git show-ref` is much slower than `git branch` and `git tag`. With this PR, the loading time of the PyTorch repo homepage drops from ~20s to 5s.

I also added the reverse logic for `GetTags` which somehow left out in #13673. The reverse logic exists in the gogit variant.

### CODE REVIEW

File: modules/git/repo_branch_nogogit.go [hunk 1]
```diff
@@ -51,32 +51,19 @@ func callShowRef(repoPath, prefix, arg string) ([]string, error) {
...
-
+		// The output of branch is simply a list:
+		// LF
 		branchName, err := bufReader.ReadString('\n')
 		if err == io.EOF {
-			// This shouldn't happen... but we'll tolerate it for the sake of peace
 			return branchNames, nil
 		}
 		if err != nil {
+			// This shouldn't happen... but we'll tolerate it for the sake of peace
```
Comment (6543) [inline modules/git/repo_branch_nogogit.go:61]: ?

**Comment by 6543** [comment]
@skyline75489 can you add unit tests?

File: modules/git/repo_branch_nogogit.go (hunk 1 above)
Comment (skyline75489) [reply modules/git/repo_branch_nogogit.go:61]: I mean the old comment  is wrong, right? The position is where the return value is. So i moved it here which makes more sense to me.

**Comment by skyline75489** [comment]
I've not yet tested this enough. On my mac the performance is not stable. I think it's because some kind of disk caching by macOS. Overall show-ref is still slower since it fetches all the hashes that end up being unused. i will test this on Windows later. 

My main concern is that i don't really know if the change is completely harmless. Is there any situation that show-ref handles but branch/tag does not.

**Comment by skyline75489** [comment]
So this is what I got from my Windows machine with SSD:

| Operation | Duration |
| ---- | ---- |
| show-ref --heads | 4.06s |
| show-ref --tag | 4.05s |
| branch --list | 0.27s |
| tag --list | 0.05s | 

This is on the same machine but with a 5400 rpm hard disk:

| Operation | Duration |
| ---- | ---- |
| show-ref --heads | 5.47s |
| show-ref --tag | 5.53s |
| branch --list | 0.54s |
| tag --list | 0.04s | 


The performance seem to be CPU-bounded rather than IO-bounded. The ~20s results was on my old MBP2015.

**Comment by 6543** [comment]
> ... that i don't really know if the change is completely harmless ...

@zeripath ping ;)

**Comment by skyline75489** [comment]
FYI, the gogit variant performs basically the same as ‘show-ref’. We are not using gogit by default, though.

**Comment by skyline75489** [comment]
@6543 So i was trying this with a smaller repo (200 branches) and found that the perf diff is not very obvious. I guess the perf diff will only emerge with a very large amount of branches/objects. I don’t I can easily add a benchmark test until I can find a way to mock a very large repo.

By the way the PyTorch repo really reveals some internal issues in gitea

**Comment by lunny** [comment]
How about to keep the function `callShowRef` but add the two new functions? So that reviews will be easier.

**Comment by skyline75489** [comment]
@lunny That’s was my initial implementation. Then I found that `callShowRef` Is not being used at all. So I deleted it. You could always use ‘Split’ mode on GitHub (click the gear button than you can choose it) which will make reviewing this diff easier. If that still bothers you, let me know and I’ll add `callShowRef` back.

File: modules/git/repo_branch_nogogit.go (hunk 1 above)
Comment (skyline75489) [reply modules/git/repo_branch_nogogit.go:61]: Or maybe we should just delete this comment. The old position just does not seem right to me.

Comment (lunny) [reply modules/git/repo_branch_nogogit.go:61]: I think it should still be the original position. It means `err == io.EOF` in fact is not an err. At a normal situation, reader shouldn't return that.

Comment (zeripath) [reply modules/git/repo_branch_nogogit.go:61]: @lunny has the correct interpretation. Please revert

**Comment by zeripath** [comment]
If we are talking about performance you must tell us what version of git you are using

**Comment by zeripath** [comment]
git branch and git show-ref do different things. I suspect that you'll  be missing other branches.

**Comment by skyline75489** [comment]
On windows I’m using `git 2.29.2`. On macOS I’m using the git provided by Apple.

I’m not sure what you mean by ‘poorly’ or ‘unpacked’. Here’s how I get the PyTorch repo:

1. First `git clone —mirror https://github.com/pytorch/pytorch.git`
2. Create the repo in my self hosted gitea instance
3. Set the remote  `git remote set-url —push origin git@mygitea/myrepo.git`.
4. Finally `git push —mirror`

Be warned that the clone & push steps take a considerable amount of time if you wanna try this yourself.

As for the behavior of `show-ref` vs `branch`, I also suspect that they have different behaviors . But after testing it with PyTorch repo, the result of `branch/tag` are identical to `show-ref`, except for the formatting, of course. Still this is my concern, too. I’m glad you brought it up.

**Review by zeripath — CHANGES_REQUESTED** [review]
The referenced issue specifically references v1.13 - please ensure that this is actually needed on master since the no-go-git PR was merged.

**Comment by zeripath** [comment]
https://github.com/go-gitea/gitea/issues/14180#issuecomment-753466851 appears to indicate that show-ref is not the problem.

File: modules/git/repo_branch_nogogit.go (hunk 1 above)
Comment (skyline75489) [reply modules/git/repo_branch_nogogit.go:61]: Am I misunderstanding this? The `err == io.EOF` seems to be the only way to break out the loop normally, but "at a normal situation, reader shouldn't return that"? How should we break out the loop, though

**Comment by skyline75489** [comment]
@zeripath  I tested this on a Linux machine and found that the performance difference between `git branch` & `git show-ref --heads` are minimum. I don't have a clue why but on macOS & Windows, the performance difference is significant. So I guess this explains why no one noticed this before since most deployment of Gitea would be Linux-based.

**Comment by skyline75489** [comment]
I'll close this for now since I can not provide a definitive conclusion that this works for every repo & machine. Let's focus on whatever hot paths that causes #14180 .

**Pull request closed without merging**